// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "google/cloud/datastore_v1/proto/entity.proto";

message Json {
  string serialized = 1;
}

message UworkerEntityWrapper {
  google.datastore.v1.Entity entity = 1;
  Json changed = 2;
}

message Input {
  optional google.datastore.v1.Entity testcase = 1;
  optional google.datastore.v1.Entity metadata = 2;
  optional string testcase_id = 3;
  optional Json uworker_env = 4;
  optional string testcase_download_url = 5;
  optional string job_type = 6;
  // uworker_io is the only module that should be using this.
  optional string uworker_output_upload_url = 7;
}

enum ErrorType {
  NO_ERROR = 0;
  ANALYZE_BUILD_SETUP = 1;
  ANALYZE_NO_CRASH = 2;
  TESTCASE_SETUP = 3;
  UNHANDLED = 4;
  VARIANT_BUILD_SETUP = 5;
}

message Output {
  optional UworkerEntityWrapper testcase = 1;
  optional UworkerEntityWrapper metadata = 2;
  optional ErrorType error = 3;
  optional Json uworker_env = 4;
  optional int64 test_timeout = 5;
  optional int64 crash_time = 6;
}